x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/MappingsMerger.java

```diff
@@ -0,0 +1,100 @@
+import org.elasticsearch.xpack.core.ClientHelper;
+import org.elasticsearch.xpack.core.ml.utils.ExceptionsHelper;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+
+import static org.elasticsearch.xpack.core.ClientHelper.ML_ORIGIN;
+
+/**
+ * Merges mappings in a best effort and naive manner.
+ * The merge will fail if there is any conflict, i.e. the mappings of a field are not exactly the same.
+ */
+public final class MappingsMerger {
+
+    private MappingsMerger() {}
+
+    public static void mergeMappings(Client client, Map<String, String> headers, String[] index,
+                                     ActionListener<ImmutableOpenMap<String, MappingMetaData>> listener) {
+        ActionListener<GetMappingsResponse> mappingsListener = ActionListener.wrap(
+            getMappingsResponse -> listener.onResponse(MappingsMerger.mergeMappings(getMappingsResponse)),
+            listener::onFailure
+        );
+
+        GetMappingsRequest getMappingsRequest = new GetMappingsRequest();
+        getMappingsRequest.indices(index);
+        ClientHelper.executeWithHeadersAsync(headers, ML_ORIGIN, client, GetMappingsAction.INSTANCE, getMappingsRequest, mappingsListener);
+    }
+
+    static ImmutableOpenMap<String, MappingMetaData> mergeMappings(GetMappingsResponse getMappingsResponse) {
+        ImmutableOpenMap<String, ImmutableOpenMap<String, MappingMetaData>> indexToMappings = getMappingsResponse.getMappings();
+
+        String type = null;
+        Map<String, Object> mergedMappings = new HashMap<>();
+
+        Iterator<ObjectObjectCursor<String, ImmutableOpenMap<String, MappingMetaData>>> iterator = indexToMappings.iterator();
+        while (iterator.hasNext()) {
+            ObjectObjectCursor<String, ImmutableOpenMap<String, MappingMetaData>> indexMappings = iterator.next();
+            Iterator<ObjectObjectCursor<String, MappingMetaData>> typeIterator = indexMappings.value.iterator();
+            while (typeIterator.hasNext()) {
+                ObjectObjectCursor<String, MappingMetaData> typeMapping = typeIterator.next();
+                if (type == null) {
+                    type = typeMapping.key;
+                } else {
+                    if (type.equals(typeMapping.key) == false) {
+                        throw ExceptionsHelper.badRequestException("source indices contain mappings for different types: [{}, {}]",
+                            type, typeMapping.key);
+                    }
+                }
+                Map<String, Object> currentMappings = typeMapping.value.getSourceAsMap();
+                if (currentMappings.containsKey("properties")) {
+
+                    @SuppressWarnings("unchecked")
+                    Map<String, Object> fieldMappings = (Map<String, Object>) currentMappings.get("properties");
+
+                    for (Map.Entry<String, Object> fieldMapping : fieldMappings.entrySet()) {
```

Comment:
Are the mappings flattened or something before they are given to this method? It seems to me that this would fail late if there were two mappings that had the same upper level type, but their nested `properties` key contained the same field names, but with different types. I know this is a "best effort" type of check.

Reply:
After chatting, I missed that the entire sub-value is compared. It should be good.